lib/vquic/curl_ngtcp2.c

```diff
@@ -434,19 +481,33 @@ static int cb_recv_stream_data(ngtcp2_conn *tconn, uint32_t flags,
   CURL_TRC_CF(data, cf, "[%" CURL_PRId64 "] read_stream(len=%zu) -> %zd",
               stream_id, buflen, nconsumed);
   if(nconsumed < 0) {
+    struct Curl_easy *cdata = CF_DATA_CURRENT(cf);
+    bool fatal = cf_ngtcp2_h3_err_is_fatal((int)nconsumed);
     /* consume all bytes */
     ngtcp2_conn_extend_max_stream_offset(tconn, stream_id, buflen);
     ngtcp2_conn_extend_max_offset(tconn, buflen);
-    if(!data || (stream && stream->reset) ||
-      NGHTTP3_ERR_H3_STREAM_CREATION_ERROR == (int)nconsumed) {
-      struct Curl_easy *cdata = CF_DATA_CURRENT(cf);
-      CURL_TRC_CF(cdata, cf, "[%" CURL_PRId64 "] discard data for stream %s",
-                  stream_id, (data && stream)? "reset" : "unknown");
+
+    if(fatal) {
+      CURL_TRC_CF(cdata, cf, "[%" CURL_PRId64 "] fatal nghttp3 error on "
+                  "nghttp3_conn_read_stream() -> %" CURL_PRId64,
+                  stream_id, (curl_int64_t)nconsumed);
+    }
+    if(!data) {
+      CURL_TRC_CF(cdata, cf, "[%" CURL_PRId64 "] discard data for unknown "
+                  "stream", stream_id);
+      return 0;
```

Comment:
Connection must be closed (return NGTCP2_ERR_CALLBACK_FAILURE) when nconsumed < 0 with no exception.

Reply:
Aye. Will change this then.

Reply:
That does not work. When I reset stream 0, and packets arrive, our callback is invoked for stream 0 and the call to `nghttp3_conn_read_stream()` returns -609. If we then return NGTCP2_ERR_CALLBACK_FAILURE the connection and all other ongoing transfers break.

Any ideas?

Reply:
I think that that is due to the misuse of nghttp3_conn_shutdown_stream_read and nghttp3_conn_close_stream at https://github.com/curl/curl/blob/e1cf21b5c2572bc05504c1b3a39d1c0620123e75/lib/vquic/curl_ngtcp2.c#L236-L238

Calling these two functions do not actually shutdown stream or close stream.   They should be called after stream is actually shutdown or closed.   I am not sure when h3_data_done is called, but could be called before actual QUIC stream is closed.  Instead of calling them, you should call ngtcp2_conn_shutdown_stream.  After that, ngtcp2 stops notifying data on the closed stream.

Reply:
Thanks for the explanation. A misunderstanding how to use the API then. Will try to kill the ngtcp2 stream then.

Reply:
Should be fixed now in latest push. Thanks for the help!